contracts/utils/Funder.sol

```diff
@@ -0,0 +1,119 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import "./SelfMulticall.sol";
+import "./interfaces/IFunder.sol";
+import "@openzeppelin/contracts/utils/cryptography/MerkleProof.sol";
+import "@openzeppelin/contracts/utils/Create2.sol";
+import "./FunderDepository.sol";
+
+contract Funder is SelfMulticall, IFunder {
+    mapping(address => mapping(bytes32 => address payable))
+        public
+        override ownerToRootToFunderDepositoryAddress;
+
+    receive() external payable {}
+
+    function deployFunderDepository(
+        address owner,
+        bytes32 root
+    ) external override returns (address payable funderDepository) {
+        // Owner allowed to be zero
+        require(root != bytes32(0), "Root zero");
+        funderDepository = payable(
+            new FunderDepository{salt: bytes32(0)}(owner, root)
+        );
+        ownerToRootToFunderDepositoryAddress[owner][root] = funderDepository;
+        // We could have not stored this and used computeFunderDepositoryAddress() on the
+        // fly whenever we needed it, but doing so requires handling the FunderDepository
+        // bytecode, which ends up being more expensive than reading a bytes32 from storage
+        emit DeployedFunderDepository(funderDepository, owner, root);
+    }
+
+    // It's a bit heavy on the calldata but I don't see a way around it
+    function fund(
+        address owner,
+        bytes32 root,
+        bytes32[] calldata proof,
+        address recipient,
+        uint256 lowThreshold,
+        uint256 highThreshold
+    ) external override {
+        require(recipient != address(0), "Recipient address zero");
+        require(
+            lowThreshold <= highThreshold,
+            "Low threshold higher than high"
+        );
+        require(highThreshold != 0, "High threshold zero");
+        // https://github.com/OpenZeppelin/merkle-tree#validating-a-proof-in-solidity
+        bytes32 leaf = keccak256(
+            bytes.concat(
+                keccak256(abi.encode(recipient, lowThreshold, highThreshold))
+            )
+        );
+        require(MerkleProof.verify(proof, root, leaf), "Invalid proof");
+        // https://en.wikipedia.org/wiki/Hysteresis#In_engineering
+        uint256 recipientBalance = recipient.balance;
+        require(recipientBalance <= lowThreshold, "Balance not low enough");
+        address payable funderDepository = ownerToRootToFunderDepositoryAddress[
+            owner
+        ][root];
+        uint256 amountNeededToTopUp;
+        unchecked {
+            amountNeededToTopUp = highThreshold - recipientBalance;
+        }
+        uint256 balance = funderDepository.balance;
+        uint256 amount = amountNeededToTopUp <= balance
+            ? amountNeededToTopUp
+            : balance;
```

Comment:
Since it will not get called for both path at once, we can save 5 gas by not initializing a new `balance` value and just calling `funderDepository.balance` on both path. 

```suggestion
        uint256 amount = amountNeededToTopUp <= funderDepository.balance
            ? amountNeededToTopUp
            : funderDepository.balance;
```

Reply:
I didn't test it but I suspect that this optimizes the case of `amountNeededToTopUp <= funderDepository.balance` at the expense of the `amountNeededToTopUp > funderDepository.balance` case. In the latter case, `funderDepository.balance` is used twice (first to check the condition and second to assign to `amount`), and assigning `funderDepository.balance` to `balance` and using that should be cheaper.
However, `amountNeededToTopUp <= funderDepository.balance` can be considered to be the happy path, so this is a viable optimization.